### PR TITLE
Add modal for inviting users to a workspace

### PR DESCRIPTION
**Pull request by Jag96**
cc @mountiny 

### Details
This PR adds the ability to invite a user to a workspace.

### Fixed Issues

Fixes https://github.com/Expensify/Expensify/issues/166040

### Tests
0. Go to expensify.com.dev (or expensify.com if testing on prod) and get the policyID of your testing account. You can find this by going to https://www.expensify.com.dev/admin_policies and clicking on one of the policies and you'll see it in the URL. For example, the one I'm using to test is `4A19DC10B34DC505`.
1. Since the workspace people list page isn't done yet, the easiest way to remove users while testing is to go to the URL mentioned above, choose a policy, choose `People` and remove that way.
2. Update [InitialPage.js](https://github.com/Expensify/Expensify.cash/blob/bb4d1364116c45f0f0f35804b6d3ff9e2df1e161/src/pages/settings/InitialPage.js#L87) and add the following object to menuItems, this will allow us to easily navigate to the modal on all platforms while testing (replace the policyID with your policyID)
    1. `{translationKey: 'common.invite', icon: Profile, action: () => { Navigation.navigate(ROUTES.getWorkspaceInviteRoute('4A19DC10B34DC505')); },},`
3. Run the app and go to settings, and click on the bottom option `Invite` (the one we added in step 2), and confirm that the modal opens with the `Invite` button disabled. Confirm that the welcomeNote placeholder value contains the policy name.
4. Open the `Application` tab of the Chrome inspector (web is the easiest way to test this) and type in `policy_4A19DC10B34DC505` in the filter bar (replace `4A19DC10B34DC505` with your policyID), confirm you see the policy object w/ the employeeList
image
5. Add a new user via the invite modal, confirm that after hitting the Invite button the modal disappears.
6. Refresh the value in the `Application` tab of the Chrome inspector, confirm that the new email is added
7. In the Chrome inspector, filter using `personalDetails`, confirm the invited user's personal details show up
8. To test the failure case, you can take one of two routes:
  1. Internal to Expensify - update the dev code of the API to throw an Exception inside the `Policy_Employees_Merge` command of the PHP api
  2. External - update the `.then` handler of `Policy_Employees_Merge` in `Policy.js` so that the `data.jsonCode === 200` check is now `data.jsonCode !== 200`
9. Invite another user via the modal, confirm that the modal closes and you see an error message
10. Check the Application tab of the Chrome inspector, confirm that the invited email doesn't show up in the list

### QA Steps
N/A, this will be more easily testable by QA in https://github.com/Expensify/Expensify/issues/166468

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### CODE REVIEW

File: src/libs/actions/Policy.js [hunk 1]
```diff
@@ -30,7 +61,74 @@ function getPolicySummaries() {
...
+        .then(() => {
+            const policyDataWithoutLogin = {};
+            policyDataWithoutLogin[key] = {
+                employeeList: _.without(allPolicies[key].employeeList, login),
+            };
+            Onyx.mergeCollection(ONYXKEYS.COLLECTION.POLICY, policyDataWithoutLogin)
+                .then(() => {
+                   console.debug('done');
+                });
+        });
```
Comment (kidroca) [inline src/libs/actions/Policy.js:105]: I find merge useful if you only have partial data for a given object - delta, and want to merge it to the object without having to retrieve it from storage first, but since you already have the full data in `allPolicies` you can use `set` to write exactly what you want

```js
    const policy = _.clone(allPolicies[key]); // make a shallow copy to preserve original data
    policy.employeeList = [...policy.employeeList, login] // concat the login

    // use set
    Onyx.set(key, policy)
        .then(() => {
            const policyDataWithoutLogin = _.clone(allPolicies[key]);
            policyDataWithoutLogin.employeeList = _.without(allPolicies[key].employeeList, login),

             Onyx.set(key, policyDataWithoutLogin )
                .then(() => {
                   console.debug('done');
                });
        });
```

An slightly shorter alternative

```js
    const policy = _.clone(allPolicies[key]);
    const policyWithLogin = { ...policy, employeeList: [...policy.employeeList, login] };

    Onyx.set(key, policyWithLogin)
        .then(() => {
             // policy is the original policy without the login and we can just set it again
             Onyx.set(key, policy)
                .then(() => {
                   console.debug('done');
                });
        });
```

Comment (kidroca) [reply src/libs/actions/Policy.js:105]: `mergeCollection` would not remove items from an array, it can only overwrite or add some, but it does not work as concat

Comment (Jag96) [reply src/libs/actions/Policy.js:105]: Awesome, thanks for having a look! I'll plan on using set then 👍

**Comment by Jag96** [comment]
This is ready for review!

File: src/libs/actions/Policy.js [hunk 2]
```diff
@@ -30,7 +75,69 @@ function getPolicySummaries() {
...
+            // Save the personalDetails for the invited user in Onyx
+            if (data.jsonCode === 200 && data.personalDetails) {
+                Onyx.merge(ONYXKEYS.PERSONAL_DETAILS, formatPersonalDetails(data.personalDetails));
+                return;
+            }
+
+            // If the operation failed, undo the optimistic addition
+            const policyDataWithoutLogin = _.clone(allPolicies[key]);
+            policyDataWithoutLogin.employeeList = _.without(allPolicies[key].employeeList, login);
+            Onyx.set(key, policyDataWithoutLogin);
```
Comment (marcaaron) [inline src/libs/actions/Policy.js:127]: NAB, if we use `Onyx.merge()` this can be shortened to something like...

```js
Onyx.merge(key, {employeeList: _.without(allPolicies[key].employeeList, login)});
```

File: src/pages/workspace/WorkspaceInvitePage.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+class WorkspaceInvitePage extends React.Component {
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            emailOrPhone: '',
+            welcomeNote: '',
+        };
+
+        this.getWelcomeNotePlaceholder = this.getWelcomeNotePlaceholder.bind(this);
```
Comment (marcaaron) [inline src/pages/workspace/WorkspaceInvitePage.js:56]: NAB, I think we do not need to bind this since `inviteUser()` already has it's `this` bound to the component and the only other usage is called in the `render()`.

File: src/pages/workspace/WorkspaceInvitePage.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+            this.props.route.params.policyID);
+        Navigation.goBack();
+    }
+
+    render() {
+        return (
+            <ScreenWrapper>
+                <HeaderWithCloseButton
+                    title={this.props.translate('workspace.invite.invitePeople')}
+                    onCloseButtonPress={() => Navigation.dismissModal()}
```
Comment (marcaaron) [inline src/pages/workspace/WorkspaceInvitePage.js:90]: NAB, 

```suggestion
                    onCloseButtonPress={Navigation.dismissModal}
```

**Review by marcaaron — DISMISSED** [review]
LGTM and overall fine work! Really only had NAB comments.

File: src/languages/es.js [hunk 5]
```diff
@@ -299,5 +300,14 @@ export default {
...
             requestCall: 'Concertar una llamada',
         },
+        invite: {
+            invitePeople: 'Invite People',
+            invitePeoplePrompt: 'Invite a colleague to your workspace.',
+            personalMessagePrompt: 'Add a Personal Message (Optional)',
+            enterEmailOrPhone: 'Email or Phone',
+            pleaseEnterValidLogin: 'Please ensure the email or phone number is valid (e.g. +15005550006).',
+            genericFailureMessage: 'An error occurred inviting the user to the workspace, please try again.',
+            welcomeNote: ({workspaceName}) => `You have been invited to the ${workspaceName} Workspace! Download the Expensify mobile App to start tracking your expenses.`,
```
[the remark this answers is not in the document]
Comment (Jag96) [reply src/languages/es.js:310]: Yeah I can run these through google translate, I didn't see a process for that but that'll be better than having them in english.

File: src/libs/actions/Policy.js [hunk 6]
```diff
@@ -1,7 +1,31 @@
...
+
+let translateLocal = (phrase, variables) => translate(CONST.DEFAULT_LOCALE, phrase, variables);
+Onyx.connect({
+    key: ONYXKEYS.PREFERRED_LOCALE,
+    callback: (preferredLocale) => {
+        if (preferredLocale) {
+            translateLocal = (phrase, variables) => translate(preferredLocale, phrase, variables);
+        }
+    },
+});
```
[the remark this answers is not in the document]
Comment (Jag96) [reply src/libs/actions/Policy.js:28]: That sounds like a good follow-up to me 👍

File: src/libs/actions/Policy.js (hunk 2 above)
Comment (Jag96) [reply src/libs/actions/Policy.js:127]: I created a thread in slack about this, I was seeing an issue where `merge` didn't overwrite the newly added value (https://expensify.slack.com/archives/C01GTK53T8Q/p1623391961201700) which is why I used set here.

File: src/libs/actions/Policy.js [hunk 7]
```diff
@@ -30,7 +75,69 @@ function getPolicySummaries() {
...
+
+    // Make the API call to merge the login into the policy
+    Policy_Employees_Merge({
+        employees: JSON.stringify([{email: login}]),
+        welcomeNote,
+        policyID,
+    })
+        .then((data) => {
+            // Save the personalDetails for the invited user in Onyx
+            if (data.jsonCode === 200 && data.personalDetails) {
```
[the remark this answers is not in the document]
Comment (Jag96) [reply src/libs/actions/Policy.js:119]: This check isn't necessary, it only really prevents us from calling `merge` unnecessarily. Even if `returnPersonalDetails` is set to false and `data.personalDetails` is undefined, it won't affect the personalDetails object (I tested). As of right now there isn't any reason that return value would contain anything other than the personalDetails, so I can remove it.

File: src/libs/actions/Policy.js [hunk 8]
```diff
@@ -30,7 +75,69 @@ function getPolicySummaries() {
...
+/**
+ * Fetches the policyList from the API and saves a simplified version in Onyx
+ */
+function getPolicyList() {
+    GetPolicyList()
+        .then((data) => {
+            if (data.jsonCode === 200) {
+                const policyDataToStore = _.reduce(data.policyList, (memo, policy) => ({
+                    ...memo,
+                    [`${ONYXKEYS.COLLECTION.POLICY}${policy.id}`]: getSimplifiedMemberList(policy),
```
[the remark this answers is not in the document]
Comment (Jag96) [reply src/libs/actions/Policy.js:87]: I updated the name of the function to `getSimplifiedEmployeeListObject`, let me know if that's more clear or if you think it'd still be better to return just the array from that function instead.

File: src/libs/actions/Policy.js [hunk 9]
```diff
@@ -30,7 +75,69 @@ function getPolicySummaries() {
...
+ * @param {String} login
+ * @param {String} welcomeNote
+ * @param {String} policyID
+ */
+function invite(login, welcomeNote, policyID) {
+    const key = `${ONYXKEYS.COLLECTION.POLICY}${policyID}`;
+
+    // Make a shallow copy to preserve original data, and concat the login
+    const policy = _.clone(allPolicies[key]);
+    policy.employeeList = [...policy.employeeList, login];
```
[the remark this answers is not in the document]
Comment (kidroca) [reply src/libs/actions/Policy.js]: No, it's a shallow copy, you'll be mutating the original `employeeList` 

```js
let original = { subArray: [1, 2] }
let cloned = _.clone(original);
cloned.subArray.push(3);

console.log(original.subArray) // [1, 2, 3]
```

Reassigning with `[...policy.employeeList, login]` won't mutate original data

File: src/libs/actions/Policy.js [hunk 10]
```diff
@@ -30,7 +75,69 @@ function getPolicySummaries() {
...
+ * @param {String} login
+ * @param {String} welcomeNote
+ * @param {String} policyID
+ */
+function invite(login, welcomeNote, policyID) {
+    const key = `${ONYXKEYS.COLLECTION.POLICY}${policyID}`;
+
+    // Make a shallow copy to preserve original data, and concat the login
+    const policy = _.clone(allPolicies[key]);
+    policy.employeeList.push(login);
```
Comment (kidroca) [inline src/libs/actions/Policy.js:106]: If it's a shallow copy - you are pushing to the original `allPolicies[key].employeeList`
It' probably won't make a difference here, but mutating your **source** data is bad

File: src/libs/actions/Policy.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (marcaaron) [reply src/libs/actions/Policy.js]: Yep I am mistaken

File: src/libs/actions/Policy.js (hunk 10 above)
Comment (marcaaron) [reply src/libs/actions/Policy.js:106]: agree it doesn't really matter in this case, but should be avoided.

File: src/libs/actions/Policy.js (hunk 2 above)
Comment (marcaaron) [reply src/libs/actions/Policy.js:127]: What's the issue exactly? I can't think of a reason why this shouldn't work as we should be replacing array values when they are on objects and not concatenating them. Curious why this is not working..

File: src/libs/actions/Policy.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (Jag96) [reply src/libs/actions/Policy.js]: Ok cool, I'll revert that, even if it doesn't matter in this case I'd rather avoid having somebody do that somewhere else where it could cause problems.

File: src/libs/actions/Policy.js (hunk 2 above)
Comment (Jag96) [reply src/libs/actions/Policy.js:127]: The issue is that calling Onyx.merge on a nested employeeList won't overwrite the value of the employeeList to remove the new login. So if `test@test.com` is set in the employeeList and we call `Onyx.merge(key, {employeeList: _.without(allPolicies[key].employeeList, login)});`, the `test@test.com` login remains in employeeList.

Comment (kidroca) [reply src/libs/actions/Policy.js:127]: > as we should be replacing array values when they are on objects

I don't know about Onyx.merge but Onyx.mergeCollection would not work like that - it will not replace the array but merge it. 
So if you have object with a nested array with 3 elements and pass a delta with nested array of 2 elements - the end result is nested  array with 3 elements

Comment (kidroca) [reply src/libs/actions/Policy.js:127]: _.merge would do the same thing

Comment (marcaaron) [reply src/libs/actions/Policy.js:127]: >I don't know about Onyx.merge but Onyx.mergeCollection would not work like that - it will not replace the array but merge it.

Well, we're not discussing `mergeCollection()` in this case, but that behavior is also not consistent with my expectations or testing and again... I wouldn't expect these arrays to get concatenated recursively...

**`_.merge()`**
Screen Shot 2021-06-15 at 12 24 58 PM

Screen Shot 2021-06-15 at 12 33 46 PM

Screen Shot 2021-06-15 at 12 41 42 PM

Comment (marcaaron) [reply src/libs/actions/Policy.js:127]: Ok I think this was a slight distraction, but pretty interesting. the default behavior of `merge()` seems like not to concat.

https://github.com/lodash/lodash/issues/2872


TL;DR it seems correct to use `Onyx.set()` here (for now) - but we should clarify when `merge()` should or should not be used somewhere. Or update it to use `_.mergeWith()` + customizer like in this example:

https://lodash.com/docs/4.17.15#mergeWith

**Review by marcaaron — APPROVED** [review]

File: src/libs/actions/Policy.js (hunk 2 above)
Comment (kidroca) [reply src/libs/actions/Policy.js:127]: > the default behavior of `merge()` seems like not to concat.

That's what I tried to say as well.

Keep in mind that thought it's possible to use `_.mergeWith` it will only work for local data - you can't istruct `AsyncStorage.merge` and `AsyncStorage.multiMerge` how to merge data - these methods are not configurable and will merge the way `_.merge` works by default - arrays are not concatenated

So if you change how `Onyx.merge` to **replace** arrays, you'd have to make the same change for `Onyx.mergeCollection`, but then you'll be forced to read the entire collection since you can't define ahead of time how data should be merged

Comment (marcaaron) [reply src/libs/actions/Policy.js:127]: Yes, I think we'd have to come up with some other solution. Left some thoughts on it in https://github.com/Expensify/react-native-onyx/issues/83 if you have any context or thoughts to add.

File: src/libs/actions/Policy.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Jag96) [reply src/libs/actions/Policy.js:28]: Created a follow-up PR here: https://github.com/Expensify/Expensify.cash/pull/3716
